Elimination of the PickleIndex from Row Test Method Signature

clrudolphi opened the pull request:
### 🤔 What's changed?

This is how we eliminate the pickleIndex parameter from the signature of generated row tests.

A hash is taken of FeatureName, ScenarioName, Example tags, and row values. This hash is inserted as a hidden tag on the pickle that is generated at compile time. These are stored as resources in the test assembly.

Generated test method is changed to calculate this hash at run-time, look for a matching pickle and use its index from the list of pickles. If a test has duplicate rows (and with the same example tags) we can't predict which matching pickle will get assigned at runtime, but I think that is acceptable as there is no discernable difference between these tests anyway.

Retries still work as pickle indices are assigned for a given row hash on a round-robin basis and can be repeated.

### ⚡️ What's your motivation? 

Removes an internal implementation detail from being visible to test frameworks and other CI tools.
This addresses the concerns raised in issue #927 

### 🏷️ What kind of change is this?

- :boom: Breaking change (incompatible changes to the API)

### ♻️ Anything particular you want feedback on?

1. This prototype uses an MD5 hash of a string (which is a concatenation of Feature Name, Scenario Name, Example Tags, and Example Row table values). The hash is stored in as a tag on the pickle. Is a hash really required here? We could simply store the concatenated string as the value of the tag. (End users never see this tag)
2. If we do wish to use a hash function, is MD5 acceptable?
3. We had previously discussed on other channels to include in the hash the Example set name and/or index. I can't think of a way of obtaining that information at run-time. Let me know if I'm missing something.
4. All the hash handling logic is located in a new static class in Reqnroll (and referenced by the Reqnroll.Generator). Should this be a dependency injected class instead?
5. Is the logic used to match a running row test with a pickle acceptable (from a thread safety perspective)?

### 📋 Checklist:


- [X] I've changed the behaviour of the code
  - [X] I have added/updated tests to cover my changes.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.
- [x] Users should know about my change
  - [x] I have added an entry to the "[vNext]" section of the [**CHANGELOG**](../blob/main/CHANGELOG.md), linking to this pull request & included my GitHub handle to the release contributors list.

----

*This text was originally taken from the [template of the Cucumber project](https://github.com/cucumber/.github/blob/main/.github/PULL_REQUEST_TEMPLATE.md), then edited by hand. [You can modify the template here.](https://github.com/reqnroll/.github/edit/main/.github/PULL_REQUEST_TEMPLATE.md)*

clrudolphi committed:
Prototype; initial commit.

obligaron commented on Reqnroll/Formatters/RuntimeSupport/TestRowPickleMapper.cs:
```diff
@@ -0,0 +1,50 @@
+        var tagsList = tags ?? Enumerable.Empty<string>();
+        var rowValuesList = rowValues ?? Enumerable.Empty<string>();
+        var v = $"{featureName}|{scenarioOutlineName}|{string.Join("|", tagsList)}|{string.Join("|", rowValuesList)}";
+        return v.GetHashCode();
```
HashCodes aren't likely to overlap but they are not unique.
Is this a problem here?

clrudolphi committed:
Fixes.

clrudolphi committed:
Adjust TUnit TestGeneratorProvider's DisplayName logic.

clrudolphi committed:
Minor cleanups

gasparnagy commented on Reqnroll/Formatters/RuntimeSupport/TestRowPickleMapper.cs:
```diff
@@ -0,0 +1,57 @@
+        var tagName = $"@RowHash_{rowHash}";
+        for (int i = 0; i < pickles.Count(); i++)
+        {
+            var pickle = pickles.ElementAt(i);
```
We potentially do multiple enumeration here. We should call a ToArray or something.

gasparnagy reviewed (commented):
I have done a first round of review, here are my comments.

I still need to think of the re-run problem that was mentioned here, but I will post about that separately.

gasparnagy committed:
Merge remote-tracking branch 'origin/main' into EliminatePickleIndexFromRowTestMethodSignature

clrudolphi commented:
> I have an idea how to make this more robust and also support re-runs.

I had not yet given retry semantics much thought for this PR yet. Now that I think about it, it will be a mess and not much that we can do about it.
The timing of when a test framework may issue a retry of a pickle is out of our control. So if we have two example table rows that are identical (say R1 and R2), the sequence of execution might be:
- R1, R2 : assigned to P1, P2 respectively
- R2, R1 : assigned to P1, P2 (not strictly correct but there is no way for anyone to know the difference)
- R1, R2, R1(retry) : assigned to P1, P2, and P1 (as we re-use from the top)
- R1, R1(retry), R2 : assigned to P1, P2, and P1 (which confuses the successful completion of the two pickles)
- R2, R1, R1(retry) : assigned to P1, P2, and P1 (again confusion)

Using a first-come, first-served approach to assigning PickleIDs to rows will result in unintuitive results in the latter two situations.

I'll think about this a bit more.

gasparnagy commented:
> Using a first-come, first-served approach to assigning PickleIDs to rows will result in unintuitive results in the latter two situations.
> 
> I'll think about this a bit more.

Sure, but please note that this only happens if the two examples are fully identical and you use such retry functionality. In that case you could anyway consider that the two examples are retries of each-other. I'm not saying that the behavior is correct, but the situation is so special, that the behavior can be "good enough".

clrudolphi committed:
Made the row hash marker prefix a constant defined in a single location.

clrudolphi commented on Reqnroll/Formatters/ExecutionTracking/FeatureExecutionTracker.cs:
```diff
@@ -79,6 +79,25 @@ public async Task ProcessEvent(FeatureStartedEvent featureStartedEvent)
                 PickleIds.Add(i.ToString(), pickle.Id);
+
+                // If the pickle has a tag that begins with "@RowHash", we need to replace the pickle with a new one without that tag
+                var tagsToRemovePrefix = "@RowHash";
```
done

clrudolphi committed:
Reverted FeatureInfo's FeatureCucumberMessages to be internal; modifying the code gen to pass the FeatureInfo to the test row mapper.

clrudolphi committed:
Refactored per review suggestions and changed manner in which we track which pickleIndex is given out to an executing test.
TestRowPickleMapper is simplified and limited to hash computation and matching/removing the PickleTag from the pickle.
Tracking of which pickleIndices have been used is movedto FeatureLevelCucumberMessages.
Marker PickleTags are removed at startup (FeatureLevelCucumberMessages constructor).
PickleIndices are assigned on a round-robin basis for a given rowHash.

clrudolphi committed:
Merge branch 'main' into EliminatePickleIndexFromRowTestMethodSignature

gasparnagy committed:
Merge branch 'main' into EliminatePickleIndexFromRowTestMethodSignature

gasparnagy committed:
fix merge conflict resolution

gasparnagy committed:
cosmetic fixes

gasparnagy committed:
refactor to calculate pickles and map together in a lazy fashion

gasparnagy committed:
Move back GetPickleIndexFromTestRow to TestRowPickleMapper to avoid having an "internal" public method on FeatureInfo

gasparnagy reviewed (approved):
I think this is good now, but since this is a kind of breaking change (as we have seen), I would feel better to put this to v4. What do you think?

clrudolphi commented:
I'll take a final look at the last set of changes. 

But, yes, agreed this should wait until v4.

clrudolphi committed:
Merge branch 'main' into EliminatePickleIndexFromRowTestMethodSignature

clrudolphi committed:
Cosmetic - updated comment in the method generator to reflect the shape of the code that is being generated.
Added Changelog entry.
